File: scripts/calculateAgaRatings.pl
#!/usr/bin/perl
##############
#
# calculateAgaRatings.pl
# Use the scores to this point in the season to compute the Aga Matrix
# AgaMatrix is defined as Colley Matrix with all IAA teams repreented by a single team and using MoV
#
# Can be called with 0 arguments (use current data)
#  or called with 2 arguments (historical data) 
#
#
# Input files: 
#   ncfScoresFile - contains results of games played up to that point in the season
#
# Output files:
#   CurrentCalculatedRatings.txt  - agaMatrix ratings of all teams to that point in season. Used to populate homepage and sort for analyze schedule  (current invocation)
#    or Week{X}CalculatedRatings.txt  - agaMatrix ratings {historical invocation}
#
##############


use strict;
use Math::MatrixReal;
use Data::Dumper;
use experimental 'smartmatch';     #gets rid of CLI warning

#print "number of CLI arguments is $#ARGV\n";


my $ncfScoresFile;  
my $fbsTeams;
my $calculatedRatingsFile;
my ($year,$week);

#If called with zero arguments use current data
if ($#ARGV == -1) {  
$year = 1900 + (localtime)[5];
if (localtime[4] lt 3) {$year = $year -1;}  #If in Jan or Feb, use last year for football season year
$ncfScoresFile = "/home/neville/cfbPlayoffPredictor/data/current/ncfScoresFile.txt";  
$fbsTeams = "/home/neville/cfbPlayoffPredictor/data/$year/fbsTeamNames.txt";
$calculatedRatingsFile = "/home/neville/cfbPlayoffPredictor/data/current/CurrentCalculatedRatings.txt";
}

#If called with two arguments use historical data
elsif ($#ARGV == 1) {  
($year,$week) = @ARGV;
if (($year < 1947) || ($year > 2050))  {
print "year should be 1947 - 2050\n";
die;
 }
if (($week < 1) || ($week > 17))  {
print "week should be between 1 - 17    -   15 reserved only for 2014 season.  Otherwise 9-14 and 17\n";
die;
 }
 
$ncfScoresFile = "/home/neville/cfbPlayoffPredictor/data/$year/week$week/Week$week"."NcfScoresFile.txt";   
$fbsTeams = "/home/neville/cfbPlayoffPredictor/data/$year/fbsTeamNames.txt";
$calculatedRatingsFile = "/home/neville/cfbPlayoffPredictor/data/$year/week$week/Week$week"."CalculatedRatings.txt";
}

#called with wrong aruments, die
else {
print "Usgae: calculateAgaRatings.pl  [year week]  \n";
print "example- calculateAgaRatings.pl 2014 9\n";
die; 
}


#Read in FBS teams for this year
my (%team,%teamH) = ();   #initialize both hases in one line
open (FBSTEAMS, "<", $fbsTeams) or die "Can't open the file $fbsTeams";
while (my $line = <FBSTEAMS> )   {
    chomp ($line);
    my ($a, $b) = split(" => ", $line);   #this will result in $key holding the name of a team (1st thing before split)
    $team{$b} = $a;
    $teamH{$a} = $b;
}     #at the conclusion of this block $team{4} will be "Florida State" and $teamH{Auburn} will be "53"
my @teams=keys(%team);                     #this creates @teams array which has elements like ("127" , "32" , "90" , "118" ,... , "34") - not sure of the value of that
my $numberOfTeams = $#teams + 1;  #$#teams starts counting at zero



my $cM;   # the Colley Matrix in line 17 of the colley matrix method
my $rCV;  # the r column-vector in line 17 of the colley matrix method
my $bCV;  # the b column-vector in line 17 of the colley matrix method

my @results;
my $resultsWinner;
my $resultsLoser;

my $teamNumberWinner;
my $teamNumberLoser;

my @teamWins;
my @teamLosses;
my %seasonWins;
my %seasonLosses;

my $useMarginOfVictory = 1;   
my $mov;
my $movFactor;
my $alpha = -0.5;     #If alpha=0 this should simplify to original colley matrix.


my @agaRating;


print "number of teams is $numberOfTeams\n";
#create the matrix



#create blank matricies
$cM = new Math::MatrixReal($numberOfTeams,$numberOfTeams);
$rCV = new Math::MatrixReal($numberOfTeams,1);    #column vector is this notation. lots of rows, 1 column
$bCV = new Math::MatrixReal($numberOfTeams,1);

#create zeros for initial wins and losses for every team
for (my $i = 1; $i<$numberOfTeams+1; $i++ ) {
 $seasonWins{$team{$i}}=0;
 $seasonLosses{$team{$i}}=0;
}

#assign the diagonal row of the CM
for (my $i = 1; $i<$#teams+2; $i++ ) {        # dont understand why +2, shuoldnt it be +1
 $cM->assign($i,$i,2);     #start with the diagonal entries at +2
}

#assign the b column vector all 1s  (this, with the diagonals at 2 corresponds to 1/2)
for (my $i = 1; $i<$#teams+2; $i++ ) {        # dont understand why +2, shuoldnt it be +1
 $bCV->assign($i,1,1);  
}


#read input data from a file.
my $scoreInput;
my $k;      #there is probably a better way to do this
open (NCFSCORESFILE, "<$ncfScoresFile") or die "$! error trying to open";     
for $scoreInput (<NCFSCORESFILE>) {
#   print $scoreInput;

($resultsWinner, $resultsLoser) = (split /:/, $scoreInput);

$scoreInput =~ m/Final.*: (.+?) (\d+) - (.+?) (\d+)/;

my $aTeamName = $1;
my $aTotal = $2;
my $hTeamName = $3;
my $hTotal = $4;


# Determine if a team is a 1AA team and assign them to the team "1AA"
if (  ($aTeamName ~~ [values %team])    )  {
#that's great, away team is FBS. Do Nothing
}
else {   #otherwise it was a 1AA team
#print "FCS team $aTeamName $aTotal - $hTeamName $hTotal (FBS) on the road\n";
$aTeamName = "1AA";
}


if (  ($hTeamName ~~ [values %team])    )  {    #Like it would ever happen -- a 1A team plays a game on the road against a 1AA team
#that's great, home team is FBS. Do Nothing
}
else {   #otherwise it was a 1AA team
#print "FCS team $hTeamName $hTotal played FBS $aTeamName $aTotal at home\n";
$hTeamName = "1AA";
}

#populate the matrix with the data
$mov=($hTotal - $aTotal);


#tiered MoV AS PER PAPER WHICH WILL BE CANON
$movFactor = 0; #standard game
if (($mov>=1) && ($mov<=2)) {$movFactor=-0.2} elsif (($mov<=-1) && ($mov>=-2)) {$movFactor=+0.2} #close game type 1
if (($mov>=25) && ($mov<=34)) {$movFactor=+0.2} elsif (($mov<=-25) && ($mov>=-34)) {$movFactor=-0.2} #blowout type 1
if ($mov>=35) {$movFactor=+0.3} elsif ($mov<=-35) {$movFactor=-0.3} #blowout type 2

#add in elements of the matrix
my $x = $cM->element($teamH{$hTeamName},$teamH{$aTeamName});   #row,column so bottom left element of the 2 teams, assuming home team is lower team number
$cM->assign($teamH{$hTeamName},$teamH{$aTeamName},$x-1+($alpha*$movFactor));
my $y = $cM->element($teamH{$aTeamName},$teamH{$hTeamName});                #top right
$cM->assign($teamH{$aTeamName},$teamH{$hTeamName},$y-1+(-$alpha*$movFactor)); 

my $d1 = $cM->element($teamH{$hTeamName},$teamH{$hTeamName});     #diagonal top-left
$cM->assign($teamH{$hTeamName},$teamH{$hTeamName},$d1+1+($alpha*$movFactor));
my $d2 = $cM->element($teamH{$aTeamName},$teamH{$aTeamName});     #diagonal bottom-right
$cM->assign($teamH{$aTeamName},$teamH{$aTeamName},$d2+1+(-$alpha*$movFactor));


my $b1 = $bCV->element($teamH{$hTeamName},1);
my $b2 = $bCV->element($teamH{$aTeamName},1);

#Figure out which team won and give it to the @results array in 1-0 format
 if ($hTotal > $aTotal)  {    #home team won
$results[$k] = "$hTeamName 1-0 $aTeamName";

$seasonWins{$hTeamName} = $seasonWins{$hTeamName} + 1;
$seasonLosses{$aTeamName}++;


$bCV->assign($teamH{$hTeamName},1,$b1+0.5);
$bCV->assign($teamH{$aTeamName},1,$b2-0.5);

}
else {    #away team won. No ties anymore...
$results[$k] = "$aTeamName 1-0 $hTeamName";
$seasonWins{$aTeamName}++;
$seasonLosses{$hTeamName}++;

$bCV->assign($teamH{$aTeamName},1,$b2+0.5);
$bCV->assign($teamH{$hTeamName},1,$b1-0.5);


}
$k++;

}

close NCFSCORESFILE;



#solve for rCV
my $dim;
my $base;
my $LRM;  # this is the LR_Matrix defined in MatrixReal and returned by the method decompose_LR

$LRM = $cM->decompose_LR();

if ( ($dim,$rCV,$base) =  $LRM->solve_LR($bCV) ) {
#print "great, it solved the matrix\n";
#Note, I don't actually have to iterate. The Matrix solution takes care of that for me !
}
else {
print "crap, there was no solution to the Colley Matrix  (This should not have happened\n)";
die;
}

#print out the team ratings
#1st split the ratings out into an array with an index for each team


for (my $i = 1; $i<$#teams+2; $i++ ) {        # dont understand why +2, shuoldnt it be +1
 
#get the team rating in human readable form
my $rCVofI = $rCV->row($i);
$rCVofI =~ /\[(.+?)\]/;
$rCVofI = $1;
$rCVofI = sprintf("%.10g", $rCVofI);
#$rCVofI = substr $rCVofI, 0, 5;     # truncate out anything under thousandnths. note, this casues a loss of precision. display only.
if ($rCVofI eq 0.5)  {     #get 0.5 entries in 0.500 format for later sorting  
 $rCVofI = "0.500";
}

$agaRating[$i] = $rCVofI;
#print "$team{$i} agaRating is $rCVofI\n";
}






#print out sorted ratings for home page and analyze schedule use
open (CALCULATEDRATINGSFILE, ">$calculatedRatingsFile") or die "$! error trying to overwrite";
my @ratings;
my @sortedRatings;

for (my $i=1; $i<$#teams+2; $i++ ) {    
push @ratings, "$agaRating[$i]:$team{$i}:$agaRating[$i]: record is $seasonWins{$team{$i}} and $seasonLosses{$team{$i}}"
}
@sortedRatings = reverse sort @ratings;
for (my $i=0; $i<$#teams+1; $i++ ) {    
$sortedRatings[$i] =~ /(.+?):(.+?):(.+?):(.+)/;   #need to greedily take the last part (no : delimiter at the end)
$sortedRatings[$i] = "$2:$3:$4";   #get rid of the rating in the front (needed earlier to sort) 
#print  "$sortedRatings[$i]\n";
print CALCULATEDRATINGSFILE "$sortedRatings[$i]\n";
}
print "calculateAgaRatings done. Output written to $calculatedRatingsFile\n";
close CALCULATEDRATINGSFILE;

File: scripts/calculateAgaRatings.py
#!/usr/bin/python3
##############
#
# calculateAgaRatings.py
# Use the scores to this point in the season to compute the Aga Matrix
# AgaMatrix is defined as Colley Matrix with all IAA teams repreented by a single team and using MoV
#
# 1 argument mandatory - model weights (eta maximizing or ATS) input as "Eta" or "ATS"
#
# Can be called with 1 argument (use current data)
#  or called with 3 arguments (historical data-  year, week) 
#
#
# Input files: 
#   ncfScoresFile - contains results of games played up to that point in the season
#
# Output files:
#   CurrentCalculatedRatings{Eta|ATS}.txt  - agaMatrix ratings of all teams to that point in season. Used to populate homepage and sort for analyze schedule  (current invocation)
#    or Week{X}CalculatedRatings{Eta|ATS}.txt  - agaMatrix ratings {historical invocation}
#
##############

from datetime import datetime
import math
import sys
import re
import numpy as np
np.set_printoptions(linewidth=320, suppress=True, precision=4, threshold=100)  # prevent scrolling on prinout of large matricies and prevent scientific notation in output

# Initialize variables
ncfScoresFile = ""
fbsTeams = ""
calculatedRatingsFile = ""
year = None
week = None
alpha = -0.5
now = datetime.now()

# If called with zero arguments, error out
if len(sys.argv) == 1:
    print("Usage: calculateAgaRatings.py Eta|ATS [year week]")
    print("Example: calculateAgaRatings.py ATS 2014 9")
    print("model weights of Eta or ATS is mandatory. Year and week are optional")
    sys.exit(1)


modelWeights = sys.argv[1]

if modelWeights not in ["Eta", "ATS"]:  
    print("Model weights must be either 'Eta' or 'ATS' only")
    sys.exit(1)


# If called with one argument, use current data
if len(sys.argv) == 2:
    year = now.year
    if now.month < 3:
        year -= 1  # If in Jan or Feb, use last year for football season year
    ncfScoresFile = f"/home/neville/cfbPlayoffPredictor/data/current/ncfScoresFile.txt"
    fbsTeams = f"/home/neville/cfbPlayoffPredictor/data/{year}/fbsTeamNames.txt"
    calculatedRatingsFile = f"/home/neville/cfbPlayoffPredictor/data/current/CurrentCalculatedRatings{modelWeights}.txt"
# If called with three arguments, use historical data
elif len(sys.argv) == 4:
    year, week = int(sys.argv[2]), int(sys.argv[3])
    if 1947 <= year <= 2050 and 1 <= week <= 17:
        ncfScoresFile = f"/home/neville/cfbPlayoffPredictor/data/{year}/week{week}/Week{week}NcfScoresFile.txt"
        fbsTeams = f"/home/neville/cfbPlayoffPredictor/data/{year}/fbsTeamNames.txt"
        calculatedRatingsFile = f"/home/neville/cfbPlayoffPredictor/data/{year}/week{week}/Week{week}CalculatedRatings{modelWeights}.txt"
    else:
        print("Usage: calculateAgaRatings.py Eta|ATS [year week]")
        print("Example: calculateAgaRatings.py Eta 2014 9")
        sys.exit(1)
else:
    print("Usage: calculateAgaRatings.py Eta|ATS [year week]")
    print("Example: calculateAgaRatings.py Eta 2014 9")
    sys.exit(1)



#ATS MoV values
def calculate_mov_factor_ATS(mov):
    if (abs(mov) == 1): movFactor = -0.9
    if (abs(mov) == 2): movFactor = -0.8
    if (abs(mov) == 3): movFactor = -0.5
    if (abs(mov) == 4): movFactor = -0.4
    if (abs(mov) == 5): movFactor = -0.35
    if (abs(mov) == 6): movFactor = -0.3
    if (abs(mov) == 7): movFactor = -0.2
    if (abs(mov) == 8): movFactor = -0.1
    if (abs(mov) == 9): movFactor = -0.05
    if (abs(mov) == 10): movFactor = -0.04
    if (abs(mov) == 11): movFactor = -0.03
    if (abs(mov) == 12): movFactor = -0.02
    if (abs(mov) == 13): movFactor = -0.01
    if (abs(mov) == 14): movFactor = 0
    if (abs(mov) == 15): movFactor = 0.02
    if (abs(mov) == 16): movFactor = 0.05
    if (abs(mov) == 17): movFactor = 0.1
    if (abs(mov) == 18): movFactor = 0.15
    if (abs(mov) == 19): movFactor = 0.16
    if (abs(mov) == 20): movFactor = 0.17
    if (abs(mov) == 21): movFactor = 0.19
    if (abs(mov) == 22): movFactor = 0.2
    if (abs(mov) == 23): movFactor = 0.21
    if (abs(mov) == 24): movFactor = 0.22
    if (abs(mov) == 25): movFactor = 0.3
    if (abs(mov) == 26): movFactor = 0.32
    if (abs(mov) == 27): movFactor = 0.34
    if (abs(mov) == 28): movFactor = 0.36
    if (abs(mov) == 29): movFactor = 0.38
    if (abs(mov) == 30): movFactor = 0.4
    if (abs(mov) == 31): movFactor = 0.42
    if (abs(mov) == 32): movFactor = 0.44
    if (abs(mov) == 33): movFactor = 0.46
    if (abs(mov) == 34): movFactor = 0.48
    if (abs(mov) >= 35): movFactor = 0.5

    return movFactor



#Eta MoV values
def calculate_mov_factor_Eta(mov):
    if 1 <= abs(mov) <= 2: movFactor = -0.2  # close win, negative movFactor
    elif 3 <= abs(mov) <= 24: movFactor = 0     # standard win, no mov weight given
    elif 25 <= abs(mov) <= 34: movFactor = 0.2   # solid win
    elif abs(mov) >= 35: movFactor = 0.3   # blowout win

    return movFactor




# Read in FBS teams for this year
team = {}
teamH = {}
with open(fbsTeams, 'r') as fbsTeamsFile:
    for index, line in enumerate(fbsTeamsFile):  # use enumerate to get 0-based indices
        a, b = line.split(' => ')    #Duke => 3 from the input file.  should redo this to just a list without => 3
        team[a.strip()] = index   #index will start at 0, so team[Duke]=2, not 3.
teams = list(team.keys())
numberOfTeams = len(teams)
#print(f"number of teams for {year} is {numberOfTeams}")

# Initialize matrices
aM = np.zeros((numberOfTeams, numberOfTeams))  # 135x135 matrix of zeros, index from 0 to 134
rCV = np.zeros((numberOfTeams,1))                  # 135x1 column vector of zeros
bCV = np.zeros((numberOfTeams,1))                  # 135x1 column vector of zeros

# Create a dictionary to store season wins and losses for each team
seasonWins = {}
seasonLosses = {}
for t in teams:
    seasonWins[t] = 0
    seasonLosses[t] = 0

# Assign the diagonal of the Aga Matrix
for i in range(numberOfTeams):
    aM[i][i] = 2

# Assign the b column vector
for i in range(numberOfTeams):
    bCV[i] = 1

# Read input data from a file
k = 0
results = []
with open(ncfScoresFile, 'r') as ncfScoresFile:
    for scoreInput in ncfScoresFile:
        weekPlayed, gameStatus, scoreLine = scoreInput.split(':')
        awayResult, homeResult = scoreLine.split(' - ')  
    
        # Modified regex to capture everything up to the last number on each side
        m = re.search(r" (.+) (\d+)$", awayResult) 
        aTeamName, aTotal = m.group(1).strip(), int(m.group(2))
    
        n = re.search(r"(.+) (\d+)$", homeResult)
        hTeamName, hTotal = n.group(1).strip(), int(n.group(2))
    
        if aTeamName not in team.keys():
            aTeamName = "1AA"
        if hTeamName not in team.keys():
            hTeamName = "1AA"

        mov = hTotal - aTotal

        if modelWeights == "Eta": movFactor = calculate_mov_factor_Eta(mov)
        elif modelWeights == "ATS": movFactor = calculate_mov_factor_ATS(mov)

        i = team[hTeamName]
        j = team[aTeamName]

        if hTotal > aTotal:
            results.append(f"{hTeamName} 1-0 {aTeamName}")
            seasonWins[hTeamName] += 1
            seasonLosses[aTeamName] += 1
            aM[i][j] = aM[i][j] - 1 + alpha * movFactor
            aM[j][i] = aM[j][i] - 1 - alpha * movFactor
            aM[i][i] = aM[i][i] + 1 + alpha * movFactor
            aM[j][j] = aM[j][j] + 1 - alpha * movFactor
            bCV[i] = bCV[i] + 0.5
            bCV[j] = bCV[j] - 0.5
        else:
            results.append(f"{aTeamName} 1-0 {hTeamName}")
            seasonWins[aTeamName] += 1
            seasonLosses[hTeamName] += 1
            aM[i][j] = aM[i][j] - 1 - alpha * movFactor
            aM[j][i] = aM[j][i] - 1 + alpha * movFactor
            aM[i][i] = aM[i][i] + 1 - alpha * movFactor
            aM[j][j] = aM[j][j] + 1 + alpha * movFactor
            bCV[j] = bCV[j] + 0.5
            bCV[i] = bCV[i] - 0.5
        k += 1


#check matrix consutructed correctly
#print(f"top 10x10 of aM is:\n{aM[0:10, 0:10]}")     
#print(f"aM is:\n{aM}\n{aM.shape}")

# Get inverse and solve for rCV
aM_inv = np.linalg.inv(aM)
rCV = aM_inv @ bCV   # @ is the matrix multiplcation operator
#print(f"rCV is:\n{rCV}\n{rCV.shape}")

# Get team ratings
agaRating = [float(f'{x:.4f}'.rstrip('0').rstrip('.')) for x in np.round(rCV, decimals=4).flatten()] #round to 4 decimals and float->string->float to get 0.6 not 0.6000

# Sort and write ratings to file
ratings = [(float(agaRating[i]), teams[i], seasonWins[teams[i]], seasonLosses[teams[i]]) 
           for i in range(len(teams))]
sortedRatings = sorted(ratings, reverse=True)

with open(calculatedRatingsFile, 'w') as outFile:
    outFile.write(f"# Aga calculated ratings with {modelWeights} maximizing weights for Week {week} year {year} - Generated {now}\n")
    for rating, team, wins, losses in sortedRatings:
        outFile.write(f"{team}:{rating}:{wins}-{losses}\n")

print(f"calculateAgaRatings with {modelWeights} model weights done for year {year} week {week}. First 3 entries are:\n1. {sortedRatings[0][1]}\t- {sortedRatings[0][0]}\n2. {sortedRatings[1][1]}\t- {sortedRatings[1][0]}\n3. {sortedRatings[2][1]}\t- {sortedRatings[2][0]}\nOutput written to", calculatedRatingsFile)
